File: ts/test/c2021_01/c2021_01_04.ts
import {expect} from 'chai';

/**
 * ### Merge Two Sorted Lists
 * {@link https://leetcode.com/explore/featured/card/january-leetcoding-challenge-2021/579/week-1-january-1st-january-7th/3592/}
 */
function mergeTwoLists(l1: ListNode | null, l2: ListNode | null): ListNode | null {
    if (!l1 && !l2) return null;
    else if (!l1) return mergeTwoLists(l2, null);
    else if (!l2) return new ListNode(l1.val, mergeTwoLists(l1.next, null));
    else if (l1.val <= l2.val) return new ListNode(l1.val, mergeTwoLists(l1.next, l2));
    else return mergeTwoLists(l2, l1);
}

class ListNode {
    val: number;
    next: ListNode | null;

    constructor(val?: number, next?: ListNode | null) {
        this.val = (val === undefined ? 0 : val);
        this.next = (next === undefined ? null : next);
    }
}

describe('c2021_01_04 Merge Two Sorted Lists', () => {
    const l = (v: number, n?: ListNode | null): ListNode | null => new ListNode(v, n);

    it('Example 1: (l1 = [1,2,4], l2 = [1,3,4]) -> [1,1,2,3,4,4]', () => {
        const l1 = l(1, l(2, l(4)));
        const l2 = l(1, l(3, l(4)));
        const e = l(1, l(1, l(2, l(3, l(4, l(4))))));

        expect(mergeTwoLists(l1, l2)).to.eql(e);
    });
    it('Example 2: (l1 = [], l2 = []) -> []', () => {
        expect(mergeTwoLists(null, null)).to.eql(null);
    });
    it('Example 3: (l1 = [], l2 = [0]) -> [0]', () => {
        expect(mergeTwoLists(null, l(0))).to.eql(l(0));
    });
});

File: src/main/scala/challenge/c2021_01/c2021_01_04.scala
package challenge.c2021_01

import org.scalatest.matchers.should.Matchers
import org.scalatest.wordspec.AnyWordSpec

/**
 * [[https://leetcode.com/explore/featured/card/january-leetcoding-challenge-2021/579/week-1-january-1st-january-7th/3592/]]
 */
class c2021_01_04 extends AnyWordSpec with Matchers {
  /**
   * === Merge Two Sorted Lists ===
   *
   * Merge two sorted linked lists and return it as a '''sorted''' list.
   * The list should be made by splicing together the nodes of the first two lists.
   *
   * '''Constraints:'''
   *  - The number of nodes in both lists is in the range `[0, 50]`.
   *  - `-100 <= Node.val <= 100`
   *  - Both `l1` and `l2` are sorted in '''non-decreasing''' order.
   */
  object Solution {
    def mergeTwoLists(l1: ListNode, l2: ListNode): ListNode = {
      @scala.annotation.tailrec
      def tailRec(l1: ListNode, l2: ListNode, guard: ListNode, prev: ListNode): ListNode = {
        if (l1 == null && l2 == null) guard.next
        else if (l1 == null || (l2 != null && l1.x > l2.x)) tailRec(l2, l1, guard, prev)
        else {
          val curr = new ListNode(l1.x)
          prev.next = curr
          tailRec(l1.next, l2, guard, curr)
        }
      }

      val guard = new ListNode(0)
      tailRec(l1, l2, guard, guard)
    }
  }

  object SolutionRec {
    def mergeTwoLists(l1: ListNode, l2: ListNode): ListNode =
      if (l1 == null) l2
      else if (l2 == null) l1
      else if (l1.x <= l2.x) new ListNode(l1.x, mergeTwoLists(l1.next, l2))
      else mergeTwoLists(l2, l1)
  }

  class ListNode(_x: Int = 0, _next: ListNode = null) {
    var next: ListNode = _next
    var x: Int = _x

    override def toString: String = s"$x,$next"
  }

  import Solution.mergeTwoLists

  private def L(x: Int, n: ListNode = null) = new ListNode(x, n)

  "Example 1: (l1 = [1,2,4], l2 = [1,3,4]) -> [1,1,2,3,4,4]" in {
    val l1 = L(1, L(2, L(4)))
    val l2 = L(1, L(3, L(4)))
    val expected = L(1, L(1, L(2, L(3, L(4, L(4))))))
    mergeTwoLists(l1, l2).toString shouldBe expected.toString
  }
  "Example 2: (l1 = [], l2 = []) -> []" in {
    mergeTwoLists(null, null) shouldBe null
  }
  "Example 3: (l1 = [], l2 = [0]) -> [0]" in {
    mergeTwoLists(null, L(0)).toString shouldBe L(0).toString
  }


}
